### PR TITLE
Serialize runtime assumptions

### PR DESCRIPTION
**Pull request by dsouzai**
# Overview
Depends on https://github.com/eclipse/omr/pull/5649

This PR adds the infrastructure to serialize/deserialize runtime assumptions. It is a port of some of the code in the `snapshot` branch.

# Design
This is the general design:

`TR_RuntimeAssumptionTable::serialize` is used to serialize runtime assumptions into a buffer. It should be called in `JitShutdown`.
`TR_RuntimeAssumptionTable::deserialize` is used to deserialize runtime assumptions from a buffer. It should be called either sometime before any of the hooks have run, or lazily as the hooks are run.

## Buffer Layout
The buffer is structured in the following way:
```
SerializedDataAtlas
Serialized assumptions for kind 1
Serialized assumptions for kind 2
...
Serialized assumptions for kind n
```
`SerializedDataAtlas` is a table of contents, defined in two new structs in `TR_RuntimeAssumptionTable`:
```
struct AssumptionCountAndSize
      {
      uint32_t _count;
      uint32_t _size;
      };

   struct SerializedDataAtlas
      {
      uint32_t _size;
      AssumptionCountAndSize _numAssumptionsPerKind[LastAssumptionKind];
      };
```
Therefore, `SerializedDataAtlas` contains the size the size of the entire buffer, as well as the number of assumptions serialized + the size of the serialized data for each runtime assumption kind. `TR_RuntimeAssumptionTable::deserialize` can use this data to deserialize the assumptions.

## Serialization
The way `TR_RuntimeAssumptionTable::serialize` serializes the assumptions is:

1. Loop through the runtime assumption kinds
2. Go through all the assumptions for that kind
3. Call the virtual method serialize on the assumption

The owning metadata is the `J9JITexceptionTable` which holds a pointer to the sentinel runtime assumption that is part of the circular linked list of assumptions that belong to a particular JIT'd body. We need to store the owning metadata so that when we add the assumption to the RAT in the restore run, we can also enqueue it into the circular linked list of assumptions for the same JIT'd body.

## Deserialization
The way `TR_RuntimeAssumptionTable::deserialize` works is by looping through all the runtime assumption kinds. For each kind (that's supported), it calls the appropriate static method `deserialize`. The reason `deserialize` is static rather than virtual is because the constructors of the runtime assumptions are protected; therefore, we can't initialize a dummy instance to call a hypothetical virtual `deserialize`.

Therefore, `deserialize` takes a pointer to the start of the section that contains the deserialized assumptions for that kind as well as the number of assumptions in that section. It then loops through the section and calls make to create the runtime assumptions and add it to the RAT.

`TR_RuntimeAssumptionTable::deserialize` also takes an optional parameter `kind`, which can be used to only create assumptions of a specific kind (rather than all kinds).

### CODE REVIEW

**Comment by dsouzai** [comment]
@0xdaryl could you please review?

File: runtime/compiler/runtime/ClassUnloadAssumption.cpp [hunk 1]
```diff
@@ -606,6 +612,58 @@ TR_RuntimeAssumptionTable::assumptionCanBeSerialized(TR_RuntimeAssumptionKind ki
...
+               atlas->_numAssumptionsPerKind[k]._size += size;
+
+               owningMetadata->runtimeAssumptionList = NULL;
+               buffer += size;
+               }
+            }
+         }
+
+      if (!assumptionCanBeSerialized((TR_RuntimeAssumptionKind)k))
+         TR_ASSERT_FATAL(atlas->_numAssumptionsPerKind[k]._count == 0, "Assumption(s) of kind %d serialized!", k);
```
Comment (dsouzai) [inline runtime/compiler/runtime/ClassUnloadAssumption.cpp:670]: I was staring at this for a while wondering how this makes sense given that we're going through all the assumptions and calling serialize on them. Then I remembered that in the `snapshot` branch, when running in snapshot mode, we restrict the JIT in a way that results only in the assumptions we expect to be serialized to get created.

This check is, I guess, a way of validating that the compiler isn't generating assumptions we're not expecting to serialize.

File: runtime/compiler/runtime/ClassUnloadAssumption.cpp [hunk 2]
```diff
@@ -580,8 +589,151 @@ int32_t TR_RuntimeAssumptionTable::countRatAssumptions()
...
+               atlas->_numAssumptionsPerKind[k]._count++;
+               atlas->_numAssumptionsPerKind[k]._size += size;
+
+               owningMetadata->runtimeAssumptionList = NULL;
+               buffer += size;
+
+               OMR::RuntimeAssumption *next = cursor->getNext();
+
+               cursor->dequeueFromListOfAssumptionsForJittedBody();
+               TR_PersistentMemory::jitPersistentFree(cursor);
```
Comment (dsouzai) [inline runtime/compiler/runtime/ClassUnloadAssumption.cpp:662]: If the plan is to continue running the JVM after taking a snpashot, this shouldn't be done.

Comment (dsouzai) [reply runtime/compiler/runtime/ClassUnloadAssumption.cpp]: Considering https://github.com/eclipse/openj9/pull/11223 I updated the PR to not free the assumptions anymore.

**Comment by dsouzai** [comment]
Force pushed again to account for https://github.com/eclipse/openj9/pull/11310

File: runtime/compiler/env/RuntimeAssumptionTable.hpp [hunk 3]
```diff
@@ -124,7 +126,58 @@ class TR_RuntimeAssumptionTable
...
+    */
+   static bool assumptionCanBeSerialized(TR_RuntimeAssumptionKind kind);
+
+   uint32_t getTotalSizeOfAssumptionsToBeSerialized() { return _totalSizeOfAssumptionsToBeSerialized; }
+
+   /**
+    * @brief Allocates a buffer and serializes Runtime Assumptions
+    *
+    * This method goes through all the runtime assumptions in the RAT that are NOT marked for detach
+    * and calls the serialize on them. It then also puts the owning metadata for each assumption.
```
Comment (0xdaryl) [inline runtime/compiler/env/RuntimeAssumptionTable.hpp:144]: Instead of "the serialize", do you mean "the assumption's serialize() method" ?

Comment (0xdaryl) [inline runtime/compiler/env/RuntimeAssumptionTable.hpp:144]: > It then also puts the owning metadata for each assumption.

Where does it "put" this?

File: runtime/compiler/runtime/ClassUnloadAssumption.cpp [hunk 4]
```diff
@@ -189,6 +189,7 @@ TR_PersistentClassInfo::removeUnloadedSubClasses()
        }
     _marked=0;
     memset(_detachPending, 0, sizeof(bool)*LastAssumptionKind);
+     _totalSizeOfAssumptionsToBeSerialized=0;
```
Comment (0xdaryl) [inline runtime/compiler/runtime/ClassUnloadAssumption.cpp:192]: Fix indentation.

File: runtime/compiler/runtime/RuntimeAssumptions.hpp [hunk 5]
```diff
@@ -299,9 +299,42 @@ class TR_UnloadedClassPicSite : public OMR::ValueModifyRuntimeAssumption
...
+    */
+   static void deserialize(TR_FrontEnd *fe, TR_PersistentMemory * pm, uint8_t *cursor, uint32_t numAssumptions);
+
    private:
    uint8_t    *_picLocation;
    uint32_t    _size;
+
+   struct SerializedData
+      {
+      uint32_t                  _size;
```
Comment (0xdaryl) [inline runtime/compiler/runtime/RuntimeAssumptions.hpp:333]: Why so much whitespace before field names?

**Review by 0xdaryl — CHANGES_REQUESTED** [review]
Minor nits only.

File: runtime/compiler/env/CHTable.cpp [hunk 6]
```diff
@@ -84,6 +84,50 @@ TR_PatchJNICallSite::dumpInfo()
    TR_VerboseLog::write(" pc=%p", _pc );
    }
 
+void
+TR_PatchJNICallSite::serialize(uint8_t *cursor, uint8_t *owningMetadata)
+   {
+   // TODO: All of this data should be convered into position independent offsets
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/env/CHTable.cpp]: Opened https://github.com/eclipse/openj9/issues/11427

**Comment by dsouzai** [comment]
@0xdaryl addressed review comments.

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by dsouzai** [comment]
Closing this, as the snapshot tech uses CRIU.

**Pull request closed without merging**